jax/api.py

```diff
@@ -89,9 +90,11 @@ def __init__(self):
 
 _thread_local_state = _ThreadLocalState()
 
-def jit(fun: Callable, static_argnums: Union[int, Iterable[int]] = (),
-        device=None, backend: Optional[str] = None,
-        donate_argnums: Union[int, Iterable[int]] = ()) -> Callable:
+def jit(fun: Callable[..., T],
```

Comment:
I don't think the return type equality is technically correct, although I understand the intent. For example, here's a case where it's sort of a lie:

`jit(lambda x: x)(7)`

The return value of the `jit`-ted function would be a `DeviceArray`, but inside the lambda `x` is going to be a `Tracer` that isn't actually a `DeviceArray`.

The same holds for other transformations.

There probably is a notion of equality that holds (i.e., equal up to certain array types being considered equivalent) but you might have trouble teaching that to a Python type checker.

It might be safest to revert the changes to this file?

Reply:
Discussed offline; this will not be visible when type-checking user code. Type checks of JIT tracing code may fail due to this in future, but in that case we can either locally disable the errors within the jit internals, or remove these annotations if they cause too much toruble.

Reply:
I've added a comment above the TypeVar to give some context to this, so that if others run into issues they will know what to do.